balancer/grpclb/grpclb_util.go

```diff
@@ -85,6 +98,11 @@ func (*lbManualResolver) Close() {}
 
 // UpdateState calls cc.UpdateState.
 func (r *lbManualResolver) UpdateState(s resolver.State) {
+	r.stateMu.Lock()
+	r.lastState = s
+	r.stateUpdateRecv = true
+	r.stateMu.Unlock()
+
 	r.ccr.UpdateState(s)
```

Comment:
Hmm... so what if the channel this is attached to went idle and then `UpdateState` is called during that time?  This will just be ignored?  Maybe that's fine?  But there's definitely a race here with leaving idleness now, with `Build` being called again.

Reply:
`ccr` here is the `resolver.ClientConn` passed to the manual resolver, and is the one implemented by the `resolver_conn_wrapper` on the `grpclb` channel. The parent channel going IDLE should not affect any of this interaction. If the `grpclb` channel went IDLE, we shouldn't have this resolver active.

Reply:
> If the grpclb channel went IDLE, we shouldn't have this resolver active.

I'm not sure what this means.  If the grpclb channel is idle (due to a bug, since we always have an active stream) then we could still try to update it while it's exiting idle, causing a race between calling the old CC's `UpdateState` and the new CC being set in `ccr`.

Reply:
We don't have this type anymore. But thinking about the case you are mentioning, let's say the following happens:
- `grpclb` channel is `IDLE` because of a bug.
- the parent `grpclb` LB policy receives a `ClientConnUpdate` and it has a new set of balancer addresses, and it calls `UpdateState` on the manual resolver
- At the same time, the `grpclb` channel is exiting IDLE.
  - This will cause a new `manual.Resolver` to be built and will update the `resolver.ClientConn` inside it.
- If the call to `UpdateState` happens first, then forwarding of that update to the old `resolver.ClientConn` within the `manual.Resolver` will have no effect. But it will still update the `lastSeenState` field. So, when the `grpclb` channel eventually exits IDLE and a new `manual.Resolver` is built, it will still send this most recent `lastSeenState` to the channel.
- If the `grpclb` channel exits IDLE first, we will see two updates pushed on to the channel. One with the previous `lastSeenState` and one with the new state pushed by the `grpclb` LB policy.

Eventually, both are consistent. Am I still missing something?